File: Data Structures & Algorithms/longest-substring-without-duplicates/submission-0.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        int left = 0;
        int maxLen = 0;

        Map<Character, Integer> lastSeenChars = new HashMap<>();

        for (int right = 0; right < s.length(); right++){
            char currentChar = s.charAt(right);

            if (lastSeenChars.containsKey(currentChar) && lastSeenChars.get(currentChar) >= left){
                left = lastSeenChars.get(currentChar) + 1; // shrinks the left window   
            }

            lastSeenChars.put(currentChar, right); // update the value
            maxLen = Math.max(maxLen, right-left+1);
        }

        return maxLen;
    }
}

File: Data Structures & Algorithms/merge-strings-alternately/submission-0.py
class Solution:
    def mergeAlternately(self, word1: str, word2: str) -> str:
        left = 0
        right = 0
        res = ""

        maxLen = max(len(word1), len(word2))

        for i in range(maxLen):
            if left < len(word1):
                res += word1[left]
                left += 1
            if right < len(word2):
                res += word2[right]
                right += 1
        print(res)
        return res